Removed domain, object vs data property, rdfs:label, Array[]-> Vehicle

mspaulik opened the pull request:
Hello @bterlson, I have implemented the following:
- Remove domain from data properties.
- Distinguish between DataProperty and ObjectProperty.
- Added "rdfs:label" .
- Changed Array[] to Vehicle in Owner Object property.
- Ordering of the file so that classes are in the beginning and properties afterwards. 3rd section will be SHACL constraints.
- SHACL part.

bterlson commented on src/index.ts:
```diff
@@ -44,18 +45,21 @@ function createRdfEmitter(program: Program) {
+  function emit() 
+  {
+    // CLASS DEF 
+    navigateProgram(program, 
```
This approach of navigating the program 3 times has a rather significant drawback I think - traversal is probably the most costly operation we're doing in this emitter, and we're traversing 3 times. Doing everything in one traversal would be more performant, but also probably more clear. You could accomplish this by creating three arrays to start - shapes, properties, and constraints (or whatever you call those sections). Then, instead of calling addQuad directly, push the arguments to the appropriate array. Then once at the end, iterate over these arrays and call addQuad.

While performance probably isn't much of a concern given the size of the cadl programs we're talking about here, it would be good to be efficient when we can.

bterlson commented on src/index.ts:
```diff
@@ -73,16 +76,37 @@ function createRdfEmitter(program: Program) {
 
           if (prop.type.kind === "Model") 
           {
+              if (checkIfDataProperty(prop.type)==false)
```
```suggestion
              if (!checkIfDataProperty(prop.type))

```

bterlson replied:
Actually you have an else block, so just swap the blocks and have `if (checkIfDataProperty(prop.type))`.

bterlson commented on src/index.ts:
```diff
@@ -132,16 +226,109 @@ function createRdfEmitter(program: Program) {
+  }
+
+  function checkIfDataProperty(model: Model) {
+    const intrinsic = getIntrinsicModelName(program, model);
```
```suggestion
    return getIntrinsicModelName(program, model) !== undefined;

```

bterlson commented on src/index.ts:
```diff
@@ -132,16 +226,109 @@ function createRdfEmitter(program: Program) {
       let ns = getNsForModel(model);
-      return ns.prefix + ":" + model.name;
+
+      if (model.name == "Array")
```
In JS, you pretty much always want to use strict equals (`===`).

bterlson commented on src/index.ts:
```diff
@@ -132,16 +226,109 @@ function createRdfEmitter(program: Program) {
+  }
+
+  function nameForModelSHACL(model: Model) {
+    const intrinsic = getIntrinsicModelName(program, model);
```
This code seems identical to the above name except the latter has `_NodeShape` appended? If so, could this be just `return nameForModel(model) + "_NodeShape"`?

mspaulik commented:
Hello @bterlson , I have addressed all your comments.

Just to explain, since I really need **writer.blank()** function to be present (and its not present in Store() which I tried to use in the first place, only addBlankNode() function exists which doesn't produce same effect), I couldn't really create 3 arrays as you suggested and simply add quads to them. So what I did instead was creating 3 different writers and iterating through all 3 of them at the very end.

However, I am getting an issue with namespacing - namespaces are written to ttl 3 times and example namespace is not picked up? I am a bit confused by the function 'getNsForModel' and didn't know how to modify it to make it work. Would you mind altering the code in this function so that it works with '3 writers' logic I am currently doing? I couldn't come with better solution due to blank() function needed.

Thanks!

mspaulik commented:
Hello @bterlson,

I am having some issues when processing this model:
**model CUSIP is string;**

Basically, I am trying to generate these RDF lines: 
**ex:CUSIP a owl:DataProperty;**
**rdfs:label "CUSIP";
rdfs:range xsd:string.**

If following the same logic as we did for others models (model Person or similar standard models) it doesn't work (note: I am now doing if/else in the beginning to check what kind of model we are processing). Basically,  the function _getIntrinsicModelName_ should return **CUSIP**, however, it returns **string**? Could you have a look? Is there an issue on CADL side with this? 

I was trying to get around the issue by calling **m.name** (which returns CUSIP). However, if we do **nn(m.name)** instead of getting **ex:CUSIP** , I am getting **\<CUSIP\>**  generated in my ttl file.

I have pushed my code so you can have a look (its marked with **TODO**).

Thank you!

bterlson commented:
> If following the same logic as we did for others models (model Person or similar standard models) it doesn't work (note: I am now doing if/else in the beginning to check what kind of model we are processing). Basically, the function getIntrinsicModelName should return CUSIP, however, it returns string? Could you have a look? Is there an issue on CADL side with this?

This behavior is by design (though confusing I admit, we are looking at cleaning it up, viz: https://github.com/microsoft/cadl/issues/1187). `getIntrinsicModelName` is telling you the type of CUSIP isn't an object but a string. You should just use the model name as the name which you tried, will look into why the output looks wonky.

bterlson commented:
The angle bracket syntax occurs when referencing a node by URI. Because CUSIP doesn't have a prefix, the writer assumes it's a URI and wraps it in angle brackets. I think the right fix is to make `nameForModel` return the actual name of the model. I've committed a change that does this.

(Also note that my editor insists on formatting, if you have editor plugins available consider installing prettier and autoformatting on save)

bterlson commented on src/index.ts:
```diff
@@ -63,7 +63,9 @@ function createRdfEmitter(program: Program) {
         // CLASS PART
         // Checks if Model is actual class (model Truck) or Model is a data property (model CUSIP is string)
-        if (m.properties.size != 0) {
+        const intrinsicName = getIntrinsicModelName(program, m);
```
The real thing you intend to test here is "is this a scalar type or a model type", and this is the recommended way to do so for now. The code you had does work but might fail in the case where a user wants an object type that happens to not have any properties.

bterlson commented on src/index.ts:
```diff
@@ -286,7 +282,7 @@ function createRdfEmitter(program: Program) {
     const intrinsic = getIntrinsicModelName(program, model);
 
-    if (!intrinsic) {
+    if (!intrinsic || intrinsic !== model.name) {
```
With this change, nameForModel will return the model name if its different from the model name, which will generally be true if you are making a new intrinsic type.

You could declare e.g. `model int32 extends Cadl.int32` which this will not work with this test, so I will ponder a better way to do this.

bterlson commented on src/index.ts:
```diff
@@ -213,65 +228,63 @@ function createRdfEmitter(program: Program) {
+                quad(
+                  nameNodeShacl,
+                  nn("sh:or"),
+                  writer.list(arr) as any // error in n3 typing, this is supported
```
I verified that this works despite the type error. The types are just wrong, unfortunately, so we have to cast to any to get around it.

mspaulik replied:
Right! I was really struggling with this part before.

bterlson commented on src/index.ts:
```diff
@@ -213,65 +228,63 @@ function createRdfEmitter(program: Program) {
+    writer.addQuads(constraintQuads);
+
+    writer.end((err, result) => {
+      result = result
```
Style points 😁 Feel free to delete if you don't like it.

bterlson commented:
I've added some commits which I think address all the above feedback/issues.

The multiple namespaces in the ttl is caused by using three writers - because the writers don't know about eachother, each one writes the namespaces out. In general using three writers is likely going to be swimming upstream as concatenating three separate .ttl files together doesn't seem necessarily safe?

I investigated just doing the "store quads in an array" approach and I think I got it working? It seems to work fine to use `writer.blank` and `writer.list` with the `quad()` factory (despite a type error in the latter case), so this approach goes that route rather than use the separate writers. If I missed a reason why this approach doesn't work though let me know!

mspaulik commented:
Thank you very much for your input! I believe we can merge it now? :)

mspaulik commented:
I have also implemented some of the in-built decorators (for models and for model props). Will need to figure out how to do implementation for custom ones next week too. Please review if all makes sense :)

bterlson commented on src/index.ts:
```diff
@@ -288,6 +382,109 @@ function createRdfEmitter(program: Program) {
+    const format = getFormat(program, m);
+    if (format) {
+      arrayQuads.push(
+        quad(object, nn("sh:format"), DataFactory.literal(format))
```
This doesn't appear to be something supported by SHACL?

mspaulik replied:
Ah yes, it would be sh:pattern

bterlson commented on src/index.ts:
```diff
@@ -288,6 +382,109 @@ function createRdfEmitter(program: Program) {
+    const minValue = getMinValue(program, m);
+    if (minValue) {
+      arrayQuads.push(
+        quad(object, nn("sh:minValue"), DataFactory.literal(minValue))
```
I think minValue and maxValue are `minInclusive` and `maxInclusive` respectively.

mspaulik commented:
Hello @bterlson , fixed both. Thanks!

bterlson commented:
Looks good, nice work!
